File: go/web-search/skill.go
// Package main implements web-search-skill — search via Tavily API.
// Every SDK must pass this test before being considered compatible with the host.
//
// @skill:id      ai.luminarys.go.web-search
// @skill:name    "Web Search"
// @skill:version 1.0.0
// @skill:desc    "Performs web search. Returns structured JSON results with advanced depth."
//
//go:generate lmsk -lang go -verbose .
package main

import (
	"encoding/json"
	"errors"

	sdk "github.com/LuminarysAI/sdk-go"
)

// Search executes a web search query via Tavily API.
// @skill:method search "Search the web and return JSON results."
// @skill:param  query required "Search query string"
// @skill:result "JSON string containing search results"
func Search(ctx *sdk.Context, query string) (string, error) {
	apiKey := sdk.GetEnv("TAVILY_API_KEY")
	if apiKey == "" {
		return "", errors.New("TAVILY_API_KEY environment variable not set")
	}

	reqJson := struct {
		Query          string `json:"query"`
		SearchDepth    string `json:"search_depth"`
		IncludeFavicon bool   `json:"include_favicon"`
	}{
		Query:          query,
		SearchDepth:    "advanced",
		IncludeFavicon: false,
	}

	reqBody, _ := json.Marshal(reqJson)

	resp, err := sdk.HttpRequest(sdk.HttpRequestOptions{
		Method: "POST",
		URL:    "https://api.tavily.com/search",
		Headers: []sdk.Header{
			{Name: "Authorization", Value: "Bearer " + apiKey},
			{Name: "Content-Type", Value: "application/json"},
		},
		Body:            reqBody,
		TimeoutMs:       0,
		MaxBytes:        0,
		FollowRedirects: false,
		UseJar:          false,
	})

	return string(resp.Body), err
}

File: rust/web-search/src/skill.rs
//! web-search-skill — web search via Tavily API.
//!
//! @skill:id      ai.luminarys.rust.web-search
//! @skill:name    "Web Search"
//! @skill:version 1.0.0
//! @skill:desc    "Performs web search. Returns structured JSON results with advanced depth."
//!
//! Build:
//!   lmsk generate -lang rust -out src .
//!   cargo build --target wasm32-wasip1 --release

use luminarys_sdk::prelude::*;

// ── Skill logic ───────────────────────────────────────────────────────────────

/// Execute a web search query via Tavily API.
///
/// @skill:method search "Search the web and return JSON results."
/// @skill:param  query required "Search query string"
/// @skill:result "JSON string containing search results"
pub fn search(_ctx: &mut Context, query: String) -> Result<String, SkillError> {
    let api_key = get_env("TAVILY_API_KEY");
    if api_key.is_empty() {
        return Err(SkillError("TAVILY_API_KEY environment variable not set".into()));
    }

    let body = serde_json::json!({
        "query": query,
        "search_depth": "advanced",
        "include_favicon": false,
    });

    let resp = http_request(HttpRequestOptions {
        method: "POST".into(),
        url: "https://api.tavily.com/search".into(),
        headers: vec![
            Header { name: "Authorization".into(), value: format!("Bearer {api_key}") },
            Header { name: "Content-Type".into(), value: "application/json".into() },
        ],
        body: body.to_string().into_bytes(),
        ..Default::default()
    })?;

    Ok(String::from_utf8_lossy(&resp.body).into_owned())
}
